src/interpolate.rs

```diff
@@ -0,0 +1,210 @@
+    }
+
+    /// Borrow the `source_frames` Interpolator from the `Converter`.
+    #[inline]
+    pub fn source(&self) -> &T {
+        &self.source
+    }
+
+    /// Mutably borrow the `source_frames` Iterator from the `Converter`.
+    #[inline]
+    pub fn source_mut(&mut self) -> &mut T {
+        &mut self.source
+    }
+
+    /// Drop `self` and return the internal `source_frames` Iterator.
+    #[inline]
+    pub fn into_source(self) -> T {
+        self.source
+    }
+}
+
+impl<T, I> Iterator for Converter<T, I> 
+    where T: Iterator,
+          <T as Iterator>::Item: Frame,
+          <<T as Iterator>::Item as Frame>::Sample: Duplex<f64>,
+          I: Interpolator<Frame=<T as Iterator>::Item>
+{
+    type Item = <T as Iterator>::Item;
+
+    #[allow(unused_variables)]
+    fn next(&mut self) -> Option<Self::Item> {
+        let Converter {
+            ref mut source, 
+            ref mut interpolator,
+            ref mut interpolation_value,
+            source_to_target_ratio
+        } = *self;
+ 
+        while *interpolation_value >= 1.0 {
+            interpolator.next_source_frame(source.next());
+            *interpolation_value -= 1.0;
+        }
+
+        let out = Some(interpolator.interpolate(*interpolation_value));
+        *interpolation_value += source_to_target_ratio;
+
+        out
+    }
+
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let len_multiplier = self.source_to_target_ratio / 1.0;
+        let (source_lower, source_upper) = self.source.size_hint();
+        let lower = (source_lower as f64 * len_multiplier) as usize;
+        let upper = source_upper.map(|upper| (upper as f64 * len_multiplier) as usize);
+        (lower, upper)
+    }
+}
+
+impl<F> Interpolator for Floor<F>
+    where F: Frame,
+          <F as Frame>::Sample: Duplex<f64>
+{
+    type Frame = F;
+
+    fn interpolate(&self, _x: f64) -> Self::Frame {
+        self.left
+    }
+
+    fn next_source_frame(&mut self, source_frame: Option<Self::Frame>) {
+        self.left = source_frame.unwrap_or(self.left);
+    }
+}
+
+impl<F> Interpolator for Linear<F>
+    where F: Frame,
+          <F as Frame>::Sample: Duplex<f64>
+{
+    type Frame = F;
+
+    /// Converts linearly from the previous value, using the next value to interpolate. It is
+    /// possible, although not advisable, to provide an x > 1.0 or < 0.0, but this will just
+    /// continue to be a linear ramp in one direction or another.
+    fn interpolate(&self, x: f64) -> Self::Frame {
+        <Self::Frame as Frame>::from_fn(|idx| {
+            let left = self.left.channel(idx).unwrap_or(&<Self::Frame as Frame>::Sample::equilibrium()).to_sample::<f64>();
+            let right = self.right.unwrap_or(self.left).channel(idx).unwrap_or(&<Self::Frame as Frame>::Sample::equilibrium()).to_sample::<f64>();
+            let diff = right - left;
+            ((diff * x) + left).to_sample::<<Self::Frame as Frame>::Sample>()
+        })
```

Comment:
Perhaps this could be a little more ergonomic/efficient using `Frame::zip_map`? Something like

```rust
let left = self.left;
let right = self.right.unwrap_or(left);
left.zip_map(right, |l, r| {
    let l_f = l.to_sample::<f64>();
    let r_f = r.to_sample::<f64>();
    let diff = r_f - l_f;
    ((diff * x) + l_f).to_sample::<<Self::Frame as Frame>::Sample>()
})
```

This should allow us to avoid calling `channel(idx).unwrap_or(etc)`, as we know that the left and right frame have the same number of channels. `zip_map` can use `unsafe` code internally to avoid this bounds check, as it asserts the `NumChannels` type is the same between both frames at compile time.